spec/unit/rust-crypto/rust-crypto.spec.ts

```diff
@@ -1395,6 +1398,261 @@ describe("RustCrypto", () => {
+                    errcode: "M_NOT_FOUND",
+                    error: "Not found",
+                },
+            });
+            expect(await rustCrypto.rehydrateDeviceIfAvailable()).toBe(false);
+
+            // create a dehydrated device
+            let dehydratedDeviceBody: any;
+            fetchMock.put("path:/_matrix/client/unstable/org.matrix.msc3814.v1/dehydrated_device", (_, opts) => {
+                dehydratedDeviceBody = JSON.parse(opts.body as string);
+                return {};
+            });
+            await rustCrypto.createAndUploadDehydratedDevice();
+
+            // rehydrate the device that we just created
+            fetchMock.get("path:/_matrix/client/unstable/org.matrix.msc3814.v1/dehydrated_device", {
+                device_id: dehydratedDeviceBody.device_id,
+                device_data: dehydratedDeviceBody.device_data,
+            });
+            const eventsResponse = jest.fn((url, opts) => {
+                // rehydrating should make two calls to the /events endpoint.
+                // The first time will return a single event, and the second
+                // time will return no events (which will signal to the
+                // rehydration function that it can stop)
+                const body = JSON.parse(opts.body as string);
+                const nextBatch = body.next_batch ?? "0";
+                const events =
+                    nextBatch === "0" ? [{ sender: testData.TEST_USER_ID, type: "m.dummy", content: {} }] : [];
+                return {
+                    events,
+                    next_batch: nextBatch + "1",
+                };
+            });
+            fetchMock.post(
+                `path:/_matrix/client/unstable/org.matrix.msc3814.v1/dehydrated_device/${encodeURIComponent(dehydratedDeviceBody.device_id)}/events`,
+                eventsResponse,
+            );
+
+            expect(await rustCrypto.rehydrateDeviceIfAvailable()).toBe(true);
+            expect(eventsResponse.mock.calls).toHaveLength(2);
+        });
+
+        it("should schedule regular creation of dehydrated devices", async () => {
+            jest.useFakeTimers({ doNotFake: ["queueMicrotask"] });
+
+            const secretStorageCallbacks = {
+                getSecretStorageKey: async (keys: any, name: string) => {
+                    return [[...Object.keys(keys.keys)][0], new Uint8Array(32)];
+                },
+            } as SecretStorageCallbacks;
+            const secretStorage = new ServerSideSecretStorageImpl(new DummyAccountDataClient(), secretStorageCallbacks);
+
+            const rustCrypto = await makeTestRustCrypto(
+                makeMatrixHttpApi(),
+                testData.TEST_USER_ID,
+                undefined,
+                secretStorage,
+            );
+            const olmMachine: OlmMachine = rustCrypto["olmMachine"];
+            rustCrypto["checkKeyBackupAndEnable"] = async () => {
+                return null;
+            };
```

Comment:
again, can we use fetch-mock to make the `/room_keys/version` request return a 404 instead of this?

Reply:
Possibly.  I just cargo-culted from a different test.